internal/servers/controller/handlers/workers/worker_service.go

```diff
@@ -469,19 +469,28 @@ func (ws *workerServiceServer) CloseConnection(ctx context.Context, req *pbs.Clo
 			ClosedReason: session.ClosedReason(v.GetReason()),
 		})
 	}
-	sessRepo, err := ws.sessionRepoFn()
+	connRepo, err := ws.connectionRepoFn()
 	if err != nil {
-		return nil, status.Errorf(codes.Internal, "error getting session repo: %v", err)
+		return nil, status.Errorf(codes.Internal, "error getting connection repo: %v", err)
 	}
 
-	closeInfos, err := sessRepo.CloseConnections(ctx, closeWiths)
+	closeInfos, err := connRepo.CloseConnections(ctx, closeWiths)
 	if err != nil {
 		return nil, err
 	}
 	if closeInfos == nil {
 		return nil, status.Error(codes.Internal, "Invalid close connection response.")
 	}
 
+	sessRepo, err := ws.sessionRepoFn()
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "error getting session repo: %v", err)
+	}
+
+	for _, v := range req.GetCloseRequestData() {
+		sessRepo.TerminateSessionIfPossible(ctx, v.GetConnectionId())
+	}
+
```

Comment:
The `Connection` in each `CloseConnectionResp` in `closeInfos` has the `SessionId`. So we should be able to get the set of unique sessions that had a connection closed. That would allow `TerminateSessionIfPossible` to be given a session id, or the set of session ids, which I think willl simplify the query a little.